File: frontend/src/services/alerts.ts
import { apiRequest } from '@/services/api'

export type AlertSeverity = 'low' | 'medium' | 'high' | 'critical'
export type AlertStatus = 'open' | 'acknowledged' | 'snoozed' | 'resolved'

export interface AlertStatusHistory {
  id: string
  alert_id: string
  from_status?: string | null
  to_status: string
  reason?: string | null
  actor_id?: string | null
  actor_name?: string | null
  metadata_json: Record<string, unknown>
  created_at: string
}

export interface AlertRecord {
  id: string
  rule_id?: string | null
  rule_key: string
  alert_type: string
  title: string
  detail: string
  severity: AlertSeverity
  status: AlertStatus
  owner_id?: string | null
  owner_name?: string | null
  owner_email?: string | null
  account_id?: string | null
  account_name?: string | null
  engagement_id?: string | null
  engagement_name?: string | null
  project_name?: string | null
  source_record_type: string
  source_record_id: string
  source_record_route?: string | null
  source_evidence_json: Record<string, unknown>[]
  previous_value_json?: Record<string, unknown> | null
  new_value_json?: Record<string, unknown> | null
  recommended_action: string
  deduplication_key: string
  first_triggered_at: string
  last_triggered_at: string
  snoozed_until?: string | null
  resolved_at?: string | null
  resolved_reason?: string | null
  created_at: string
  updated_at: string
  status_history: AlertStatusHistory[]
}

export interface AlertPage {
  items: AlertRecord[]
  total: number
  page: number
  page_size: number
  pages: number
}

export interface AlertRule {
  id: string
  rule_key: string
  name: string
  description: string
  alert_type: string
  source_type: string
  threshold_value: number
  threshold_unit: string
  severity: AlertSeverity
  snooze_days: number
  recipient_policy: 'source_owner_first' | 'account_owner_first' | string
  escalation_enabled: boolean
  is_active: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface AlertRulePreview {
  rule_id: string
  rule_key: string
  total_matches: number
  sample: {
    account_id?: string | null
    account_name?: string | null
    engagement_id?: string | null
    engagement_name?: string | null
    source_record_type: string
    source_record_id: string
    title: string
    detail: string
    severity: AlertSeverity
    evidence: Record<string, unknown>[]
  }[]
}

export interface AlertEvaluationResult {
  evaluated: number
  matched: number
  created: number
  updated: number
  resolved: number
  reactivated: number
  notifications_created: number
  worker_run_id?: string | null
}

function query(params: Record<string, string | number | boolean | undefined | null>) {
  const search = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') search.set(key, String(value))
  })
  const value = search.toString()
  return value ? `?${value}` : ''
}

export function getAlerts(token: string, params: Record<string, string | number | undefined | null> = {}) {
  return apiRequest<AlertPage>(`/api/alerts${query(params)}`, { token })
}

export function getAlert(token: string, alertId: string) {
  return apiRequest<AlertRecord>(`/api/alerts/${alertId}`, { token })
}

export function updateAlertStatus(token: string, alertId: string, payload: { status: AlertStatus; reason?: string; snoozed_until?: string | null }) {
  return apiRequest<AlertRecord>(`/api/alerts/${alertId}/status`, { token, method: 'PATCH', body: JSON.stringify(payload) })
}

export function evaluateAlerts(token: string, payload: { scope: 'all' | 'account'; account_id?: string; rule_id?: string }) {
  return apiRequest<AlertEvaluationResult>('/api/alerts/evaluate', { token, method: 'POST', body: JSON.stringify(payload) })
}

export function getAlertRules(token: string) {
  return apiRequest<AlertRule[]>('/api/admin/alert-rules', { token })
}

export function updateAlertRule(token: string, ruleId: string, payload: Partial<Pick<AlertRule, 'is_active' | 'threshold_value' | 'severity' | 'snooze_days' | 'recipient_policy' | 'escalation_enabled'>>) {
  return apiRequest<AlertRule>(`/api/admin/alert-rules/${ruleId}`, { token, method: 'PATCH', body: JSON.stringify(payload) })
}

export function previewAlertRule(token: string, ruleId: string) {
  return apiRequest<AlertRulePreview>(`/api/admin/alert-rules/${ruleId}/preview`, { token, method: 'POST' })
}

File: backend/app/repositories/alerts.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import String, case, cast, func, or_, select
from sqlalchemy.orm import Session, selectinload

from app.models import Alert, AlertRule, AlertStatusHistory, ScheduledWorkerRun


class AlertRepository:
    def __init__(self, db: Session) -> None:
        self.db = db

    def list_rules(self) -> list[AlertRule]:
        return list(self.db.scalars(select(AlertRule).order_by(AlertRule.sort_order, AlertRule.name)))

    def list_active_rules(self, *, rule_id: str | None = None) -> list[AlertRule]:
        conditions = [AlertRule.is_active.is_(True)]
        if rule_id:
            conditions.append(AlertRule.id == rule_id)
        return list(self.db.scalars(select(AlertRule).where(*conditions).order_by(AlertRule.sort_order, AlertRule.name)))

    def get_rule(self, rule_id: str) -> AlertRule | None:
        return self.db.get(AlertRule, rule_id)

    def get_rule_by_key(self, rule_key: str) -> AlertRule | None:
        return self.db.scalar(select(AlertRule).where(AlertRule.rule_key == rule_key))

    def save_rule(self, rule: AlertRule) -> AlertRule:
        self.db.add(rule)
        self.db.flush()
        return rule

    def get_alert(self, alert_id: str) -> Alert | None:
        return self.db.scalar(
            select(Alert)
            .where(Alert.id == alert_id)
            .options(selectinload(Alert.status_history), selectinload(Alert.account), selectinload(Alert.owner))
        )

    def find_active_alert_by_deduplication_key(self, deduplication_key: str) -> Alert | None:
        return self.db.scalar(
            select(Alert)
            .where(Alert.deduplication_key == deduplication_key, Alert.status.in_(("open", "acknowledged", "snoozed")))
            .order_by(Alert.created_at.desc())
            .limit(1)
        )

    def list_current_alerts_for_scope(self, *, rule_ids: list[str], account_id: str | None = None) -> list[Alert]:
        conditions = [Alert.rule_id.in_(rule_ids) if rule_ids else False, Alert.status.in_(("open", "acknowledged", "snoozed"))]
        if account_id:
            conditions.append(Alert.account_id == account_id)
        return list(self.db.scalars(select(Alert).where(*conditions)))

    def list_alerts(
        self,
        *,
        account_ids: list[str] | None = None,
        status_filter: str | None = None,
        severity: str | None = None,
        alert_type: str | None = None,
        account_id: str | None = None,
        owner_id: str | None = None,
        source_type: str | None = None,
        search: str | None = None,
        now: datetime | None = None,
        page: int = 1,
        page_size: int = 25,
    ) -> tuple[list[Alert], int]:
        conditions = []
        if account_ids is not None:
            conditions.append(Alert.account_id.in_(account_ids) if account_ids else False)
        if status_filter == "active":
            conditions.append(Alert.status.in_(("open", "acknowledged", "snoozed")))
            if now is not None:
                conditions.append(or_(Alert.status != "snoozed", Alert.snoozed_until.is_(None), Alert.snoozed_until <= now))
        elif status_filter:
            conditions.append(Alert.status == status_filter)
        if severity:
            conditions.append(Alert.severity == severity)
        if alert_type:
            conditions.append(Alert.alert_type == alert_type)
        if account_id:
            conditions.append(Alert.account_id == account_id)
        if owner_id:
            conditions.append(Alert.owner_id == owner_id)
        if source_type:
            conditions.append(Alert.source_record_type == source_type)
        if search and search.strip():
            term = f"%{search.strip()}%"
            conditions.append(
                or_(
                    Alert.title.ilike(term),
                    Alert.detail.ilike(term),
                    Alert.rule_key.ilike(term),
                    Alert.account_name.ilike(term),
                    Alert.owner_name.ilike(term),
                    Alert.recommended_action.ilike(term),
                    cast(Alert.source_evidence_json, String).ilike(term),
                )
            )
        total = self.db.scalar(select(func.count(Alert.id)).where(*conditions)) or 0
        items = list(
            self.db.scalars(
                select(Alert)
                .where(*conditions)
                .options(selectinload(Alert.status_history), selectinload(Alert.account), selectinload(Alert.owner))
                .order_by(
                    Alert.status.asc(),
                    case(
                        (Alert.severity == "critical", 4),
                        (Alert.severity == "high", 3),
                        (Alert.severity == "medium", 2),
                        (Alert.severity == "low", 1),
                        else_=0,
                    ).desc(),
                    Alert.last_triggered_at.desc(),
                    Alert.created_at.desc(),
                )
                .offset((page - 1) * page_size)
                .limit(page_size)
            )
        )
        return items, total

    def save_alert(self, alert: Alert) -> Alert:
        self.db.add(alert)
        self.db.flush()
        return alert

    def save_status_history(self, history: AlertStatusHistory) -> AlertStatusHistory:
        self.db.add(history)
        self.db.flush()
        return history

    def save_worker_run(self, run: ScheduledWorkerRun) -> ScheduledWorkerRun:
        self.db.add(run)
        self.db.flush()
        return run

    def commit(self) -> None:
        self.db.commit()
